### DISCUSSION TITLE
Ubuntu/Docker | Storage - Git/Local setup permission denied..

### DISCUSSION THREAD

**Discussion by tracer8** [requarks/wiki#3841] [Error / Bug Report]
Hi, 
After install wiki.js to my ubuntu server use docker. I try to config my storage. And nothing success, it always trigger errors:
**With git option:**
+ It can't even read a file from path i give, file is created and available, i try to open it with Ubuntu command, it can still open.
![PicPick_025](https://user-images.githubusercontent.com/15725821/114871911-fcf6aa80-9e23-11eb-9344-6f7f09cdd8da.jpg)

+ Try to use contents option instead of path, same result:
![PicPick_026](https://user-images.githubusercontent.com/15725821/114872068-257ea480-9e24-11eb-9186-10c8de35ced2.jpg)

**With local option:**
+ It said can't create a directory by permisison denied? 
![PicPick_024](https://user-images.githubusercontent.com/15725821/114872280-5c54ba80-9e24-11eb-8005-30d5f5d20628.jpg)

I think Wiki Docker have some problem with permission on my Ubuntu server, so it don't have any permission to create and access files?
Please help me finger it out this issue. Thanks!

**Comment by tracer8** [comment]
May be my install location cause this error?
Install location: /etc/wiki

**Comment by Spivoran** [comment]
Hello,
I had some trouble with permissions on the default folder for local storage too. I'm using the Digital Ocean one-click droplet, which is an Ubuntu instance running WIKI.JS containers. As a Docker novice, I had some conceptual issues understanding the local storage option and where the files/folders were actually located. In case you are having a similar issue, here's how I understand how to make it work. Of course I'm sure there are other approaches that may utilize better, more accepted practices, but this did the trick for me:

- The operative question for Local Storage is Local to what? It's important to note that the folder for local storage will reside on the wiki's docker container. If you want to explore the structure of the installation, SSH into your Ubuntu server as root and poke around.
- `docker ps` will list all of the containers running on the server. Mine has three, `requarks/wiki:2`, `requarks/wiki-update-companion:latest`, and `postgress:11`.
- If you copy the container ID of the `requarks/wiki:2` container, you can then use it to open a BASH prompt inside the container.
- `docker exec -it <CONTAINER ID> bash`
- From there you can explore the container's file system and attempt to make a new directory for backups. When I did this, I noticed I only had permissions to modify the `/wiki` folder. So, instead of changing any permissions, I created a backup folder there:
- `mkdir /wiki/wiki-backup`, then `exit` back to the Ubuntu server prompt. Note: the WIKI.JS admin should make its own directory, making the preceding steps unnecessary, but I wanted to share how to access the container's file system.
- Then, switching back to the wiki admin page, I was able to successfully create local storage in that location and perform a `Dump All Contents To Disk` operation.
- This should get you to a working local storage option. Just as a way to help others, I will share the other steps to copy the files from the local folder to my laptop for archival purposes:
- Back on the SSH prompt for the Ubuntu server, copy the files from the docker container to the Ubuntu server's file system:
- `docker cp <containerid>:/wiki/wiki-backup ~/`
- The files will now be reachable at `/root/wiki-backup` on the Ubuntu server.
- Now `exit` to the command prompt for your local machine and use rsync to pull the folder from the server:
- `rsync -a root@<YOUR SERVER HOSTNAME>:/root/wiki-backup <LOCAL PATH TO DESTINATION>`

**Comment by fnafox** [comment]
Like Spivoran said, the context of what "local" means is important. If your wiki is running in a container, then "Local" means files that are in the container. If you want to use an external key (such as one that is on your host machine) you will need to bind the file onto the docker container using the docker flag `--mount` or `-v`:

```
docker run -d \
-p 8080:3000 \
--name wiki \
--restart unless-stopped \
-v /etc/wiki/gitlab.pem:/wiki/gitlab.pem \
ghcr.io/requarks/wiki:2
```
Here the file /etc/wiki/gitlab.pem is mounted to the location /wiki/gitlab.pem in the container, and then you can reference that in the git configuration.
